src/charm.py

```diff
@@ -149,25 +159,28 @@ def _get_interfaces(self):
             raise ErrorWithStatus(err, BlockedStatus)
         return interfaces
 
-    def _get_relational_db_data(
-        self,
-    ):
-        relational_db = self.model.relations["relational-db"]
-        if len(relational_db) > 1:
-            raise ErrorWithStatus(
-                f"Too many mysql relations. Found {len(relational_db)}, expected 1", BlockedStatus
-            )
-
-        try:
-            relational_db = relational_db[0]
-            db_unit = list(relational_db.units)[0]
-            relational_db = relational_db.data[db_unit]
-            relational_db["database"]
-            return relational_db
-        except IndexError:
-            raise ErrorWithStatus("Waiting for relational-db relation data", WaitingStatus)
-        except KeyError:
-            raise ErrorWithStatus("Missing data in relational-db relation", WaitingStatus)
+    def _get_relational_db_data(self) -> dict:
+        mysql_relation = self.model.get_relation("relational-db")
+
+        # Raise exception and stop execution if the relational-db relation is not established
+        if not mysql_relation:
+            raise ErrorWithStatus("Please add relation to the database", BlockedStatus)
+
+        data = self.database.fetch_relation_data()
+        self.logger.debug("Got following database data: %s", data)
+        for val in data.values():
+            if not val:
+                continue
+            self.logger.info("New mysql database endpoint is %s", val["endpoints"])
+            host, port = val["endpoints"].split(":")
+            db_data = {
+                "host": host,
+                "port": port,
+                "username": val["username"],
+                "password": val["password"],
+            }
+            return db_data
```

Comment:
Can we also add `raise ErrorWithStatus` for the case where we do not find required attributes?

Reply:
Added handling for `KeyError` exception + added unit test.